# Exploring eBay Car Sales Data

## Table of Contents

1. [**Introduction**](#1)
    - Project Description
    - Data Description
2. [**Acquiring and Loading Data**](#2)
	- Importing Libraries and Notebook Setup
    - Loading Data
    - Basic Data Exploration
    - Areas to Fix
3. [**Data Proprocessing**](#3)
4. [**Exploratory Data Analysis**](#4)
5. [**Conclusion**](#5)
    - Insights
    - Suggestions
    - Possible Next Steps
6. [**Epilogue**](#6) 
    - References
    - Versioning

---

# 1

## Introduction

![eBay Kleinanzeighen]("/Users/aleciaduncan/Documents/GitHub/Data-Analysis/eBay_Klein/ebay_ger.png")
### Project Description

**Goal/Purpose:** 

What is this project about? What is the the goal/purpose of this project? Why is it important for someone to read this notebook?
This project aims to look at information about used car listings fromn a classifieds listing on the German eBay website. 
<p>&nbsp;</p>

**Questions to be Answered:**

- 
- Question 2
- Question 3...

<p>&nbsp;</p>

**Assumptions/Methodology/Scope:** 

This data is originally in German, therefore it may be necessary to translate some values to English for clarity. 

<p>&nbsp;</p>

### Data Description

**Content:** 

This dataset describes used car listings from eBay Kleinanzeigen (a classifieds section of the German eBay website.) 

<p>&nbsp;</p>

**Description of Attributes:** 

Here you can describe what each column represents.
| Column  | Description |
| :------ | :---------- |
| dateCrawled | When this ad was first crawled. All field-values are taken from this date. |
| name | Name of the car.|
| seller | Whether the seller is private or a dealer.|
| offerType | The type of listing sell the car. |
| abtest | Whether the listing is included in an A/B test. |
| vehicleType | The vehicle type. |
| yearOfRegistration | The year in which the car was first registered. |
| gearbox | The transmission type. |
| powerPS | The power of the car in PS. |
| model | The car model name. |
| odometer | How many kilometers the car has driven. |
| monthOfRegistration | The month in which the car was first registered. |
| fuelType | What type of fuel the car uses.|
| brand | The brand of the car. |
| notRepairedDamage | If the car has any damage which is not yet repaired.|
| dateCreated | The date on whihc the eBay listing was created.|
| nrOfPictures | The number of pictures in the ad.|
| postalCode | The postal code for the location of the vehicle.|
| lastSeenOnline | When the crawler last saw this ad online.|


<p>&nbsp;</p>

**Acknowledgements:** 
This dataset was originally scraped and uploaded to Kaggle by user [ogresleka](https://www.kaggle.com/orgesleka). The data is no longer available on Kaggle but can be found [here](https://data.world/data-society/used-cars-data). For purposes of this project, the dataset has been reduced by DataQuest to a sampling of 50,000 data points. The dataset originally uploaded to Kaggle was quite clean, so this dataset was also "dirtied" to help more closely resemble a scraped dataset. 

---

# 2

## Acquiring and Loading Data
### Importing Libraries and Notebook Setup

In [400]:
# Ignore warnings if needed
import warnings
warnings.filterwarnings('ignore')

# Data manipulation
import datetime
import numpy as np
import pandas as pd

# Pandas settings
pd.options.display.max_columns = None
pd.options.display.max_colwidth = 60
pd.options.display.float_format = '{:,.3f}'.format

### Loading Data

In [401]:
# Load DataFrame
file = 'autos.csv'
autos = pd.read_csv(file, encoding = "Latin-1")

### Basic Data Exploration

#### Number of Rows and Columns

In [402]:
# Show rows and columns count
print(f"Rows count: {autos.shape[0]}\nColumns count: {autos.shape[1]}")

Rows count: 50000
Columns count: 20


#### Display First and Last Rows

In [403]:
# Look at first 5 rows
autos.head()

dateCrawled  \
0  2016-03-26 17:47:46   
1  2016-04-04 13:38:56   
2  2016-03-26 18:57:24   
3  2016-03-12 16:58:10   
4  2016-04-01 14:38:50   

                                                          name  seller  \
0                             Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1                   BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2                                   Volkswagen_Golf_1.6_United  privat   
3          Smart_smart_fortwo_coupe_softouch/F1/Klima/Panorama  privat   
4  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepflegt.mit_Klim...  privat   

  offerType   price   abtest vehicleType  yearOfRegistration    gearbox  \
0   Angebot  $5,000  control         bus                2004    manuell   
1   Angebot  $8,500  control   limousine                1997  automatik   
2   Angebot  $8,990     test   limousine                2009    manuell   
3   Angebot  $4,350  control  kleinwagen                2007  automatik   
4   Angebot  $1,350     test       kombi                2003    manuell   

   powerPS   model   odometer  monthOfRegistration fuelType       brand  \
0      158  andere  150,000km                    3      lpg     peugeot   
1      286     7er  150,000km                    6   benzin         bmw   
2      102    golf   70,000km                    7   benzin  volkswagen   
3       71  fortwo   70,000km                    6   benzin       smart   
4        0   focus  150,000km                    7   benzin        ford   

  notRepairedDamage          dateCreated  nrOfPictures  postalCode  \
0              nein  2016-03-26 00:00:00             0       79588   
1              nein  2016-04-04 00:00:00             0       71034   
2              nein  2016-03-26 00:00:00             0       35394   
3              nein  2016-03-12 00:00:00             0       33729   
4              nein  2016-04-01 00:00:00             0       39218   

              lastSeen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

In [404]:
# Look at last 5 rows
autos.tail()

dateCrawled  \
49995  2016-03-27 14:38:19   
49996  2016-03-28 10:50:25   
49997  2016-04-02 14:44:48   
49998  2016-03-08 19:25:42   
49999  2016-03-14 00:42:12   

                                                              name  seller  \
49995             Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon  privat   
49996  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+Reifen_neu_!!  privat   
49997                               Fiat_500_C_1.2_Dualogic_Lounge  privat   
49998                           Audi_A3_2.0_TDI_Sportback_Ambition  privat   
49999                                          Opel_Vectra_1.6_16V  privat   

      offerType    price   abtest vehicleType  yearOfRegistration    gearbox  \
49995   Angebot  $24,900  control   limousine                2011  automatik   
49996   Angebot   $1,980  control      cabrio                1996    manuell   
49997   Angebot  $13,200     test      cabrio                2014  automatik   
49998   Angebot  $22,900  control       kombi                2013    manuell   
49999   Angebot   $1,250  control   limousine                1996    manuell   

       powerPS   model   odometer  monthOfRegistration fuelType brand  \
49995      239      q5  100,000km                    1   diesel  audi   
49996       75   astra  150,000km                    5   benzin  opel   
49997       69     500    5,000km                   11   benzin  fiat   
49998      150      a3   40,000km                   11   diesel  audi   
49999      101  vectra  150,000km                    1   benzin  opel   

      notRepairedDamage          dateCreated  nrOfPictures  postalCode  \
49995              nein  2016-03-27 00:00:00             0       82131   
49996              nein  2016-03-28 00:00:00             0       44807   
49997              nein  2016-04-02 00:00:00             0       73430   
49998              nein  2016-03-08 00:00:00             0       35683   
49999              nein  2016-03-13 00:00:00             0       45897   

                  lastSeen  
49995  2016-04-01 13:47:40  
49996  2016-04-02 14:18:02  
49997  2016-04-04 11:47:27  
49998  2016-04-05 16:45:07  
49999  2016-04-06 21:18:48

#### Check Data Types

In [405]:
# Show data types
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

- all column names use a modified CamelCase, or a mixedCase, naming convention and will be changed to snake_case.
- a few column names can be renamed in a way that is just as informative but more concise.
- `yearOfRegistration`, `powerPS`, `monthOfRegistration`, `nrOfPictures`, and `postalCode` are **integers**.
- the other 15 columns are **strings**.
- `vehicleType`, `fuelType`, `model`, `gearbox` and `vehicleType` all contain null values and will need additional exploration.



#### Check Missing Data

In [406]:
# Print percentage of missing values
missing_percent = autos.isna().mean().sort_values(ascending=False)
print('---- Percentage of Missing Values (%) -----')
if missing_percent.sum():
    print(missing_percent[missing_percent > 0] * 100)
else:
    print(None)

---- Percentage of Missing Values (%) -----
notRepairedDamage   19.658
vehicleType         10.190
fuelType             8.964
model                5.516
gearbox              5.360
dtype: float64


#### Check for Duplicate Rows

In [407]:
# Show number of duplicated rows
print(f"No. of entirely duplicated rows: {autos.duplicated().sum()}")

# Show duplicated rows
autos[autos.duplicated()]

No. of entirely duplicated rows: 0


Empty DataFrame
Columns: [dateCrawled, name, seller, offerType, price, abtest, vehicleType, yearOfRegistration, gearbox, powerPS, model, odometer, monthOfRegistration, fuelType, brand, notRepairedDamage, dateCreated, nrOfPictures, postalCode, lastSeen]
Index: []

#### Check Uniqueness of Data

In [408]:
# Print the number of unique values
num_unique = autos.nunique().sort_values()
print('---- Number of Unique Values -----')
print(num_unique)

---- Number of Unique Values -----
nrOfPictures               1
notRepairedDamage          2
abtest                     2
gearbox                    2
seller                     2
offerType                  2
fuelType                   7
vehicleType                8
odometer                  13
monthOfRegistration       13
brand                     40
dateCreated               76
yearOfRegistration        97
model                    245
powerPS                  448
price                   2357
postalCode              7014
name                   38754
lastSeen               39481
dateCrawled            48213
dtype: int64


#### Check Data Range

In [409]:
# Print summary statistics
autos.describe(include='all')

dateCrawled         name  seller offerType  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicleType  yearOfRegistration  gearbox    powerPS  model   odometer  \
count        44905          50,000.000    47320 50,000.000  47242      50000   
unique           8                 NaN        2        NaN    245         13   
top      limousine                 NaN  manuell        NaN   golf  150,000km   
freq         12859                 NaN    36993        NaN   4024      32424   
mean           NaN           2,005.073      NaN    116.356    NaN        NaN   
std            NaN             105.713      NaN    209.217    NaN        NaN   
min            NaN           1,000.000      NaN      0.000    NaN        NaN   
25%            NaN           1,999.000      NaN     70.000    NaN        NaN   
50%            NaN           2,003.000      NaN    105.000    NaN        NaN   
75%            NaN           2,008.000      NaN    150.000    NaN        NaN   
max            NaN           9,999.000      NaN 17,700.000    NaN        NaN   

        monthOfRegistration fuelType       brand notRepairedDamage  \
count            50,000.000    45518       50000             40171   
unique                  NaN        7          40                 2   
top                     NaN   benzin  volkswagen              nein   
freq                    NaN    30107       10687             35232   
mean                  5.723      NaN         NaN               NaN   
std                   3.712      NaN         NaN               NaN   
min                   0.000      NaN         NaN               NaN   
25%                   3.000      NaN         NaN               NaN   
50%                   6.000      NaN         NaN               NaN   
75%                   9.000      NaN         NaN               NaN   
max                  12.000      NaN         NaN               NaN   

                dateCreated  nrOfPictures  postalCode             lastSeen  
count                 50000    50,000.000  50,000.000                50000  
unique                   76           NaN         NaN                39481  
top     2016-04-03 00:00:00           NaN         NaN  2016-04-07 06:17:27  
freq                   1946           NaN         NaN                    8  
mean                    NaN         0.000  50,813.627                  NaN  
std                     NaN         0.000  25,779.748                  NaN  
min                     NaN         0.000   1,067.000                  NaN  
25%                     NaN         0.000  30,451.000                  NaN  
50%                     NaN         0.000  49,577.000                  NaN  
75%                     NaN         0.000  71,540.000                  NaN  
max                     NaN         0.000  99,998.000                  NaN

### Areas to Fix
**Data Types**
- `odometer`, `price` columns need to be converted from strings to integers and floats respectively
- `dateCrawled`, `yearOfRegistration`, `monthOfRegistration`, and `last_seen` should all be converted to datetime datatypes
- `postalCode` is listed as an integer and calculations were made accordingly, because of the type of data a postal code is this may be better suited as a string datatype

**Missing Data**
- `unrepaired_damage` is missing 19.658% of it's data, need to check to see if missing values are indicative of $0 unrepaird damage or some other issue
- `vehicle_type` is missing 10.190% of it's data, these rows may need to be removed 
- `fuel_type` 8.964%
- `model` 5.516%
- `gearbox` 5.360%

**Duplicate Rows**
- no duplicated rows

**Uniqueness of Data**
- `numPictures` only has one unique value and can be dropped
- `seller` and `offer_type` columns only have 2 unique values, when looking at the frequency of the top value it can be seen that 49,999 of the 50,000 are the same. both of these columns are redundant and can be dropped, will confirm with a redundancy 
- `abtest`, `gearbox` and `notRepairedDamage` only have 2 unique values but each with significant amounts - additional exploration necessary


**Data Range**
- `yearOfRegistration` appears to have minimum and maximum values exceeding reasonable years for these listings
- `powerPS` has a minimum value of 0 and a maximum value of 17,700, but the values at the 25th and 75th percentile indicate the column as a whole trends far below the maximum - additional exploration necessary

---

# 3

## Data Preprocessing

### Rename Columns

In [410]:
# Rename columns
new_columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_pictures', 'postal_code',
       'last_seen']
autos.columns = new_columns

In [411]:
# Verify columns are renamed
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

### Drop Redundant Columns

In [412]:
# Check the proportion of the most frequent value in each column
print('---- Frequency of the Mode (%) -----')
mode_dict = {col: (autos[col].value_counts().iat[0] / autos[col].size * 100) for col in autos.columns}
mode_series = pd.Series(mode_dict)
mode_series.sort_values(ascending=False)

---- Frequency of the Mode (%) -----


num_pictures         100.000
seller                99.998
offer_type            99.998
gearbox               73.986
unrepaired_damage     70.464
odometer              64.848
fuel_type             60.214
abtest                51.512
vehicle_type          25.718
brand                 21.374
powerPS               11.000
registration_month    10.150
model                  8.048
registration_year      6.708
ad_created             3.892
price                  2.842
postal_code            0.218
name                   0.156
last_seen              0.016
date_crawled           0.006
dtype: float64

In [413]:
# Show the value frequency of each column greater than the mode's threshold
threshold = 80          
for col in mode_series[mode_series > threshold].index:
    print(autos[col].value_counts(dropna=False))
    print()

privat        49999
gewerblich        1
Name: seller, dtype: int64

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

0    50000
Name: num_pictures, dtype: int64



In [414]:
# Drop columns 
cols_to_drop = ["seller", "offer_type", "num_pictures"]
autos.drop(columns=cols_to_drop, axis=1, inplace=True)

In [415]:
# Verify columns dropped
assert all(col not in autos.columns for col in cols_to_drop)

### Changing Data Types

In [416]:
# Convert columns to the right data types

autos["odometer"] = autos["odometer"].str.replace("[^0-9.]", "", regex = True).astype(int)
autos["price"] = autos["price"].str.replace("[^0-9.]", "", regex = True).astype(float)

# Update column name to better reflect data as converted
autos.rename(columns={"odometer" : "odometer_km"}, inplace = True)


### Handling Missing Values

### Handling Unreasonable Data Ranges

**Column: `odometer_km`**

In [417]:
autos["odometer_km"].nunique()

13

In [418]:
autos["odometer_km"].describe()

count    50,000.000
mean    125,732.700
std      40,042.212
min       5,000.000
25%     125,000.000
50%     150,000.000
75%     150,000.000
max     150,000.000
Name: odometer_km, dtype: float64

In [419]:
autos["odometer_km"].value_counts().sort_index(ascending=False)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

In [420]:
z_scores = {}
for mileage in autos["odometer_km"].unique():
    z_score = ((mileage - 125732.7) / 40042.212)
    if (z_score >= 3) | (z_score <= -3):
        z_scores[mileage] = z_score
print(z_scores)

{5000: -3.015135627372434}


In [421]:
#Drop all rows with an outlier value 
autos = autos[autos.odometer_km != 5000]

#Verify columns dropped equate to the expected amount of dropped columns
new_len = len(autos)
print(new_len, (50000 - new_len == 967))


49033 True


**Column: `price`**

In [422]:
autos["price"].nunique()

2323

In [423]:
autos["price"].describe()

count       49,033.000
mean         9,809.323
std        485,497.272
min              0.000
25%          1,150.000
50%          2,980.000
75%          7,200.000
max     99,999,999.000
Name: price, dtype: float64

In [424]:
autos["price"].value_counts().sort_index(ascending=True).head(10)

0.000     1291
1.000      135
2.000        1
3.000        1
5.000        2
10.000       5
11.000       2
12.000       1
13.000       1
14.000       1
Name: price, dtype: int64

In [425]:
autos["price"].value_counts().sort_index(ascending=False).head(15)

99,999,999.000    1
27,322,222.000    1
12,345,678.000    3
11,111,111.000    2
10,000,000.000    1
1,300,000.000     1
1,234,566.000     1
999,999.000       2
999,990.000       1
345,000.000       1
299,000.000       1
259,000.000       1
220,000.000       1
197,000.000       1
190,000.000       1
Name: price, dtype: int64

(Note: the top 9 values here represent the top 13 rows of prices and appear to include listings over $990,000. In a dataset of 50,000, these items appear to be outliers, skew the data, and are small enough that removing them will be insignificant for the purposes of this project.)

In [426]:
autos = autos[autos.price < 990000]
print(len(autos), ((49033 - len(autos)) == 13))

49020 True


In [427]:
autos["price"].describe()

count    49,020.000
mean      5,688.790
std       8,411.410
min           0.000
25%       1,150.000
50%       2,950.000
75%       7,200.000
max     345,000.000
Name: price, dtype: float64

In [429]:
# Calculate z-scores for values to determine outliers (z-score of +/- 3)
z_scores = {}

for price in autos["price"].unique():
    z_score = ((price - 5688.79)/8411.41)
    if (z_score >= 3) | (z_score <= -3):
        z_scores[price] = z_score

In [430]:
# Retrieve the price threshold for which values exceeded a z-score of 3
z_score_prices = z_scores.keys()
sorted_prices = sorted(z_score_prices)
threshold = sorted_prices[0]

print(threshold)

30933.0


In [431]:
#Drop all rows above the price threshold
autos = autos[autos.price < threshold]

#Verify rows were dropped
print(len(autos))

48306


### Feature Engineering / Transformation

---

# 4

## Exploratory Data Analysis

Here is where your analysis begins. You can add different sections based on your project goals.

### Exploring `odometer_km`

In [ ]:
# Code and visualization

**Observations**
- Ob 1
- Ob 2
- Ob 3

### Exploring `prices`

---

# 5

## Conclusion

### Insights 
State the insights/outcomes of your project or notebook.

### Suggestions

Make suggestions based on insights.

### Possible Next Steps
Areas to expand on:
- (if there is any)

---

# 6

## Epilogue

### References

This is how we use inline citation[<sup id="fn1-back">[1]</sup>](#fn1).

[<span id="fn1">1.</span>](#fn1-back) _subject (date)._ Title. Available at: https://website.com (Accessed: Date). 

> Use [https://www.citethisforme.com/](https://www.citethisforme.com/) to create citations.

### Versioning
Notebook and insights by (author).
- Version: 1.0.0
- Date: 